### ISSUE TITLE
Generated VAPI depenencies no longer passed to valac when building an executable

### ISSUE THREAD

**Issue by jtojnar** [mesonbuild/meson#7628]
**Describe the bug**
Building GIMP master branch with Meson 0.55.0 or master fails to build on building Vala `executable` due to underlinking VAPI dependencies:

```
[4/1371] Compiling Vala source libgimp/gimp-3.0.vapi libgimp/gimp-ui-3.0.vapi ../plug-ins/goat-exercises/goat-exercise-vala.vala
FAILED: plug-ins/goat-exercises/goat-exercise-vala.p/goat-exercise-vala.c 
valac -C --debug --debug --disable-since-check --pkg gegl-0.4 --pkg gtk+-3.0 --color=always --directory plug-ins/goat-exercises/goat-exercise-vala.p --basedir ../plug-ins/goat-exercises libgimp/gimp-3.0.vapi libgimp/gimp-ui-3.0.vapi ../plug-ins/goat-exercises/goat-exercise-vala.vala
gegl-0.4.vapi:56.71-56.74: error: The symbol `Babl' could not be found
		public void set_color_from_pixel (Gegl.Rectangle rect, void* pixel, Babl.Object pixel_format);
		                                                                    ^^^^
gegl-0.4.vapi:114.18-114.21: error: The symbol `Babl' could not be found
		public unowned Babl.Object? get_format ();
		               ^^^^
...
```

* Downstream issue: https://gitlab.gnome.org/GNOME/gimp/-/issues/5407

**To Reproduce**
* VAPI: https://gitlab.gnome.org/GNOME/gimp/-/blob/acf50009f8cec8f9e18bc64ca3b6b76929c4835d/libgimp/meson.build#L459-474
* Executable: https://gitlab.gnome.org/GNOME/gimp/-/blob/acf50009f8cec8f9e18bc64ca3b6b76929c4835d/plug-ins/goat-exercises/meson.build#L41-51

**Expected behavior**
It should build like it does with Meson 0.54.3, ~~currently bisecting~~.

It appears that the regression was caused by 7195c0ee9c3df499a2e9015e44ddcd2718d03c8b.

**system parameters**
* native build
* Nix unstable, Fedora Rawhide…
* Python 3.8.5
* Meson 0.55.0, fbc6f50acc
* ninja 1.10.0

**Comment by oleavr** [comment]
@jtojnar The bug I fixed was hiding an existing bug elsewhere. Meson was previously letting targets mutate things that were supposed to be constant, so one target could add flags used by all other targets that made the same query after it.

**Comment by lazka** [comment]
I'm hitting the same bug when building gegl in MSYS2.

**Comment by nirbheek** [comment]
I can reproduce this with GIMP on Fedora.

**Comment by nirbheek** [comment]
So, I don't think this is a bug in meson. It's one of those rare times when two bugs happen to cancel each other out. In 0.54, due to a bug, `find_library()` was having side-effects, so when gimp did `valac.find_library('babl-0.1')`, that was adding `--pkg babl-0.1` to `coredata.external_args`, and then was getting added to _all_ vala targets.

However, what GIMP needs to do is to add the babl dependency that it found to the vala targets that need it. Not doing that is a bug.

These two bugs were cancelling each other out. The fix here is in GIMP:

```patch
diff --git a/meson.build b/meson.build
index a4958d2ec3..4c766ce9d8 100644
--- a/meson.build
+++ b/meson.build
@@ -748,15 +748,8 @@ endif

 have_vala = add_languages('vala', required: get_option('vala-plugins'))
 if have_vala
-  babl = declare_dependency(
-    dependencies: [
-      babl,
-      meson.get_compiler('vala').find_library('babl-0.1'),
-    ]
-  )
-
-  # TODO: remove this once we release 3.0
   valac = meson.get_compiler('vala')
+  # TODO: remove this once we release 3.0
   if valac.version().version_compare('>= 0.31.1')
     add_project_arguments('--disable-since-check', language: 'vala')
   endif
diff --git a/plug-ins/goat-exercises/meson.build b/plug-ins/goat-exercises/meson.build
index 982400778d..399de45f63 100644
--- a/plug-ins/goat-exercises/meson.build
+++ b/plug-ins/goat-exercises/meson.build
@@ -42,7 +42,11 @@ if have_vala
     'goat-exercise-vala.vala',
     include_directories: [ rootInclude, ],
     dependencies: [
-      libgimp_vapi, libgimpui_vapi, gtk3, gegl, math,
+      libgimp_vapi, libgimpui_vapi, gtk3, gegl,
+      # pkgconfig file is babl.pc and vala dependency is babl-0.1.vapi, i.e.
+      # they have different names, so we can't reuse the babl dependency found
+      # with the C compiler like we do for other deps.
+      valac.find_library('babl-0.1'),
     ],
     c_args: [
       '-DGETTEXT_PACKAGE="@0@"'.format(gettext_package),
```

I will open an MR on GIMP for this.

@lazka I cannot reproduce the gegl issue you're seeing with introspection on Linux. It is also unrelated to this because gegl does not use any vala code. That's something else.

**Comment by lazka** [comment]
> @lazka I cannot reproduce the gegl issue you're seeing with introspection on Linux. It is also unrelated to this because gegl does not use any vala code. That's something else.

I got the same build error as OP though (with MSYS2). But thanks for looking into it. Maybe something has changed in gegl master already, I'll need to recheck. Feel free to ignore my issue for continuing here.

**Comment by nirbheek** [comment]
> I got the same build error as OP though (with MSYS2)

I saw your comment here: https://github.com/msys2/MINGW-packages/pull/6924#issuecomment-686988336 and the error is different. That one is not coming from `valac`. It does have the same root cause though: Gegl uses Babl symbols in its API.

**Comment by Jehan** [comment]
I answered on GIMP side, but let's make it too! I don't really know much about how this VAPI dependency works, but if one lib has a dependency on another, then it should bring it itself somehow. This is also how `pkg-config` works with the `Requires:` keyword (so that the right flags and libs are automatically used in the build). Nowadays one cannot ask a project to follow up all dependency changes of each of its first-level dependencies.

So unless I am mistaken, I am guessing this problem is either that GEGL does not properly defines babl as a dependency to VAPI (though @jtojnar thinks that's what [this piece of build code](https://gitlab.gnome.org/GNOME/gegl/-/blob/69a4d515ac4a0154f42d992b7520e3bce42142bf/gegl/meson.build#L136-145) is for), or else it's indeed a bug in meson which doesn't propagate dependencies correctly.

**Comment by nirbheek** [comment]
Unlike with C, where we need to resolve dependencies by looking at pkgconfig files, we do not need to pre-resolve dependencies with Vala. The compiler will do that for us by looking at the corresponding `.deps` file.

I tested with Gegl master + meson 0.54.3 + vapigen 0.48.8 and the .deps file is generated correctly. However, the distro-shipped `gegl-0.4.deps` file is empty, which causes valac to throw those errors. This is almost certainly not a meson bug.

**Comment by nirbheek** [comment]
> However, the distro-shipped `gegl-0.4.deps` file is empty

Looks like that was a gegl bug, which was fixed in https://gitlab.gnome.org/GNOME/gegl/-/merge_requests/83. Closing.
